mijia/src/lib.rs

```diff
@@ -1,137 +1,107 @@
+            //     )], outer_sig: Signature("a{sv}") })
+            // instead.
+            let service_data: HashMap<String, _> = device_properties
+                .get("ServiceData")?
+                // Variant(...)
+                .as_iter()?
+                .next()?
+                // InternalDict(...)
+                .as_iter()?
+                .tuples::<(_, _)>()
+                .filter_map(|(k, v)| Some((k.as_str()?.into(), v)))
+                .collect();
+
             if service_data.contains_key(MIJIA_SERVICE_DATA_UUID) {
-                sensors.push(device);
+                Some(SensorProps {
+                    object_path: path.to_string(),
+                    mac_address,
+                })
+            } else {
+                None
             }
-        }
-    }
-
-    sensors
-}
-
-pub fn print_sensors(sensors: &[BluetoothDevice], sensor_names: &HashMap<String, String>) {
-    println!("{} sensors:", sensors.len());
-    for device in sensors {
-        let mac_address = device.get_address().unwrap();
-        // TODO: Find a less ugly way to do this.
-        let empty = "".to_string();
-        let name = sensor_names.get(&mac_address).unwrap_or(&empty);
-        println!(
-            "{}: {:?}, {} services, {} service data, '{}'",
-            mac_address,
-            device.get_name(),
-            device.get_gatt_services().map_or(0, |s| s.len()),
-            device.get_service_data().map_or(0, |s| s.len()),
-            name
-        );
-    }
+        })
+        .collect();
+    Ok(sensors)
 }
 
-pub fn connect_sensor<'a>(sensor: &BluetoothDevice<'a>) -> bool {
-    if let Err(e) = sensor.connect(CONNECT_TIMEOUT_MS) {
-        println!("Failed to connect {:?}: {:?}", sensor.get_id(), e);
-        false
-    } else {
-        println!("Connected to {:?}", sensor.get_id());
-        true
-    }
-}
-
-pub fn connect_sensors<'a>(sensors: &'a [BluetoothDevice<'a>]) -> Vec<BluetoothDevice<'a>> {
-    let mut connected_sensors = vec![];
-    for device in sensors {
-        if connect_sensor(device) {
-            connected_sensors.push(device.clone());
-        }
-    }
-
-    println!("Connected to {} sensors.", connected_sensors.len());
-
-    connected_sensors
-}
-
-pub fn start_notify_sensor<'a>(
-    bt_session: &'a BluetoothSession,
-    connected_sensor: &BluetoothDevice<'a>,
-) -> Result<(), Box<dyn Error>> {
-    let temp_humidity = BluetoothGATTCharacteristic::new(
-        bt_session,
-        connected_sensor.get_id() + SERVICE_CHARACTERISTIC_PATH,
+pub async fn start_notify_sensor(
+    bt_session: &MijiaSession,
+    device_path: &str,
+) -> Result<(), anyhow::Error> {
+    let temp_humidity_path: String = device_path.to_string() + SERVICE_CHARACTERISTIC_PATH;
+    let temp_humidity = dbus::nonblock::Proxy::new(
+        "org.bluez",
+        temp_humidity_path,
+        Duration::from_secs(30),
```

Comment:
Put this as a constant up the top of the file. I assume it is a timeout of some sort?